### PR TITLE
Support 32 bit RGBA sprite sequences

### PR DESCRIPTION
**Pull request by pchote**
Last year, #15194 unified Indexed vs RGBA sprites at the shader level and removed the performance cliff caused by switching between them. This PR takes another step on the slow-burning renderer rewrite by carrying the unification through the `Sequence` and `SpriteLoader` code.

Modders should not get too excited just yet: all the caveats of not having a palette still apply, so RGBA sprites won't yet work with player colours, target flashes, map lighting, alpha/shadows, and probably more. I expect that this won't be properly usable until we implement vertex-based color tinting/alpha, mask or chroma-key based player colors, and #3692.

The testcase demonstrates @mabulsoud's radar dome from #13245 ingame.

A more in-depth testcase is available on my [d2k-r16](https://github.com/pchote/OpenRA/tree/d2k-r16) branch, which switches D2k to use the higher quality R16 sprites (which you will need to find and install manually).

### CODE REVIEW

**Comment by dragunoff** [comment]
I've been using this PR for a while and successfully loading 32-bit sprites but I noticed that R and B channels are swapped. If I for example pass this in my sprite's image data: `(uint)Color.Red.ToArgb()` it would render as blue.

File: OpenRA.Mods.Common/SpriteLoaders/PngSheetLoader.cs [hunk 1]
```diff
@@ -69,24 +66,29 @@ public bool TryParseSprite(Stream s, out ISpriteFrame[] frames, out TypeDictiona
 			{
 				var frameStart = frameRegions[i].X + frameRegions[i].Y * png.Width;
```
Comment (dragunoff) [inline OpenRA.Mods.Common/SpriteLoaders/PngSheetLoader.cs]: I think these should be multiplied by `pixelLength` to get the offset for the frame:
```csharp
var frameStart = frameRegions[i].X * pixelLength + frameRegions[i].Y * pixelLength * png.Width;
```

**Comment by dragunoff** [comment]
Another weird thing I noticed is that some images are cut off (missing one row or column of pixels), the TD icon for example:
изображение

**Comment by abcdefg30** [comment]
Are you certain this is a regression from this PR? We had issues with such missing lines before as well.

**Comment by dragunoff** [comment]
> 
> 
> Are you certain this is a regression from this PR? We had issues with such missing lines before as well.

No, I'm not sure about that. But I think it tends to happen with 32-bit sprites.

**Comment by pchote** [comment]
@dragunoff can you provide your testcases?

**Comment by dragunoff** [comment]
@pchote Please have a look at https://github.com/dragunoff/OpenRA/commit/87131231388884d86a7b3e689c66935b602fb71c

I've added some simple color bars that render fine a single frame but are all jumbled up when sliced into frames. I have not tested the sprites in-game, only in the asset browser. I hope the metadata in the frames is correct. The TD logo also renders with the first row of the image missing (screenshot in an earlier message).

**Comment by dragunoff** [comment]
The problem I illustrated above is actually caused by the wrong frame offset calculation that I already mentioned in https://github.com/OpenRA/OpenRA/pull/16885#discussion_r320494401.

After fixing that the problem goes away and frames render correctly (see 
https://github.com/dragunoff/OpenRA/commit/ad7f468d486d4cfba7959719dd5b7e8d4fd543ae). 

I will check again how sprites render on my downstream branch and if I am still able to reproduce the problem.

**Comment by dragunoff** [comment]
OK, this is happening **only on Windows and only in the asset browser**. It also affects 8-bit PNGs and my custom sprite loaders so it's probably a different issue altogether (something with the asset browser maybe?). I discovered this by testing on an Ubuntu machine where everything looks fine both in-game and in the asset browser. Here's what I see in the asset browser on Windows 7:
vQxkBqADr0

**Comment by pchote** [comment]
Rebased and fixed the frame offset issue. I still need to look into the reversed color channels.

**Comment by pchote** [comment]
Updated. The color channel issue was a silly oversight: OpenRA internally uses BGRA channel order, not RGBA. Renamed `SpriteFrameType.RGBA` to `SpriteFrameType.BGRA` to account for this.

The cursor changes turned into a bigger rework, so I will file these shortly as a followup PR.

Would you mind giving this another look @dragunoff?

**Comment by dragunoff** [comment]
@pchote Seems fine on my downstream branch and the testcases that I posted earlier look good too. The asset browser glitch on Windows 7 is still there.

**Comment by pchote** [comment]
Do you have Windows configured with >100% DPI scaling and `Graphics.DisableWindowsDPIScaling` disabled in your settings.yaml? This could easily cause both the asset browser glitch here and the cursor distortion in #17517 without the patches on my `render-wip` branch.

**Comment by dragunoff** [comment]
@pchote How do I check for these settings?

**Comment by pchote** [comment]
Open `settings.yaml` in your support dir and delete the `DisableWindowsDPIScaling` line if it exists under the `Graphics` section.

**Comment by dragunoff** [comment]
No such line there and I don't think I have messed with the DPI scaling in Windows:
изображение

**Comment by dragunoff** [comment]
@pchote Does this PR add support for 32-bit sprite annotations? I'm talking about the move flash in particular. I tried but it doesn't seem to work.

**Comment by dragunoff** [comment]
@pchote Please ignore my previous comment - move flash works beatifuly, I had a mistake in the yaml.

**Comment by pchote** [comment]
Rebased on #17517 and added the (now much smaller) RGBA cursor commits here.

I have not yet fixed the color flipping issue (caused by incorrectly caching data) raised by @dragunoff in #17517.

**Comment by pchote** [comment]
Updated. Duplicated cursors should now work correctly.

I can't reproduce the asset browser png issue in my Windows 10 VM, so I don't know what is going on there.

**Comment by pchote** [comment]
Rebased.

**Comment by pchote** [comment]
I would really like to remove the requirement to restart after changing the "Increase Command Cursor Size" setting for #17345, but doing this cleanly requires merging `HardwareCursor` and `SoftwareCursor`, which requires this PR.

There should be nothing controversial, or even particularly complicated here. Can we please try and make some progress here?

File: OpenRA.Game/Graphics/Util.cs [hunk 2]
```diff
@@ -61,23 +61,57 @@ public static void FastCreateQuad(Vertex[] vertices, float3 a, float3 b, float3
...
+
+								data[(y + j) * destStride + x + i] = PremultiplyAlpha(cc).ToArgb();
+							}
+						}
+					}
 				}
+			}
+			else
+			{
+				var data = dest.Sheet.GetData();
```
Comment (abcdefg30) [inline OpenRA.Game/Graphics/Util.cs]: This seems to be redundant to `destData`?

File: OpenRA.Mods.Common/UtilityCommands/DumpSequenceSheetsCommand.cs [hunk 3]
```diff
@@ -44,14 +44,19 @@ void IUtilityCommand.Run(Utility utility, string[] args)
...
 			foreach (var s in sb.AllSheets)
 			{
 				var max = s == sb.Current ? (int)sb.CurrentChannel + 1 : 4;
 				for (var i = 0; i < max; i++)
 					s.AsPng((TextureChannel)ChannelMasks[i], palette).Save("{0}.png".F(count++));
 			}
 
+			sb = sequences.SpriteCache.SheetBuilders[SpriteFrameType.BGRA];
+			foreach (var s in sb.AllSheets)
+					s.AsPng().Save("{0}.png".F(count++));
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/UtilityCommands/DumpSequenceSheetsCommand.cs]: The indentation here looks wrong.

File: OpenRA.Mods.D2k/SpriteLoaders/R8Loader.cs [hunk 4]
```diff
@@ -22,6 +22,8 @@ public class R8Loader : ISpriteLoader
 	{
 		class R8Frame : ISpriteFrame
 		{
+			public SpriteFrameType Type { get; set; }
+
```
Comment (reaperrr) [inline OpenRA.Mods.D2k/SpriteLoaders/R8Loader.cs]: Minor consistency nit (not a blocker): This is the only loader you added an empty line to. Might want to drop this one while adressing @abcdefg30's points.

**Review by reaperrr — DISMISSED** [review]
Ingame looks good, didn't spot any code issues.

**Comment by pchote** [comment]
Fixed, rebased, and removed testcase.

File: OpenRA.Game/Graphics/Util.cs [hunk 5]
```diff
@@ -1,6 +1,6 @@
 #region Copyright & License Information
 /*
- * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
```
Comment (abcdefg30) [inline OpenRA.Game/Graphics/Util.cs]: Time travel.

**Comment by pchote** [comment]
Fixed.

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/0da929264049105ad228082970171d2bec475f43...a228e94ffcd53ab4be155b5b86f160916761a0e6)

**Pull request merged**